beacon-chain/db/filesystem/blob.go

```diff
@@ -0,0 +1,176 @@
+	ethpb "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	"github.com/spf13/afero"
+)
+
+var (
+	errIndexOutOfBounds = errors.New("blob index in file name > MaxBlobsPerBlock")
+)
+
+const (
+	sszExt       = "ssz"
+	partExt      = "part"
+	blobLockPath = "blob.lock"
+)
+
+// NewBlobStorage creates a new instance of the BlobStorage object. Note that the implementation of BlobStorage may
+// attempt to hold a file lock to guarantee exclusive control of the blob storage directory, so this should only be
+// initialized once per beacon node.
+func NewBlobStorage(base string) (*BlobStorage, error) {
+	base = path.Clean(base)
+	if err := os.MkdirAll(base, 0755); err != nil {
+		return nil, err
+	}
+	fs := afero.NewBasePathFs(afero.NewOsFs(), base)
+	return &BlobStorage{fs: fs}, nil
+}
+
+// BlobStorage is the concrete implementation of the filesystem backend for saving and retrieving BlobSidecars.
+type BlobStorage struct {
+	fs afero.Fs
+}
+
+// Save saves blobs given a list of sidecars.
+func (bs *BlobStorage) Save(sidecar blocks.VerifiedROBlob) error {
+	fname := namerForSidecar(sidecar)
+	sszPath := fname.ssz()
+	exists, err := afero.Exists(bs.fs, sszPath)
+	if err != nil {
+		return err
+	}
+	if exists {
+		// TODO: should it be an error to save a blob that already exists?
+		return nil
+	}
+
+	// Serialize the ethpb.BlobSidecar to binary data using SSZ.
+	sidecarData, err := ssz.MarshalSSZ(sidecar)
+	if err != nil {
+		return errors.Wrap(err, "failed to serialize sidecar data")
+	}
+	if err := bs.fs.Mkdir(fname.dir(), 0755); err != nil {
+		return err
+	}
+	partPath := fname.partial()
+	// Create a partial file and write the serialized data to it.
+	partialFile, err := bs.fs.Create(partPath)
+	if err != nil {
+		return errors.Wrap(err, "failed to create partial file")
+	}
+
+	_, err = partialFile.Write(sidecarData)
+	if err != nil {
+		closeErr := partialFile.Close()
+		if closeErr != nil {
+			return closeErr
+		}
+		return errors.Wrap(err, "failed to write to partial file")
+	}
+	err = partialFile.Close()
+	if err != nil {
+		return err
+	}
+
+	// Atomically rename the partial file to its final name.
+	err = bs.fs.Rename(partPath, sszPath)
+	if err != nil {
+		return errors.Wrap(err, "failed to rename partial file to final name")
+	}
+	return nil
+}
+
+// Get retrieves a single BlobSidecar by its root and index.
+// Since BlobStorage only writes blobs that have undergone full verification, the return
+// value is always a VerifiedROBlob.
+func (bs *BlobStorage) Get(root [32]byte, idx uint64) (blocks.VerifiedROBlob, error) {
+	expected := blobNamer{root: root, index: idx}
+	encoded, err := afero.ReadFile(bs.fs, expected.ssz())
+	var v blocks.VerifiedROBlob
+	if err != nil {
+		return v, err
+	}
+	s := &ethpb.BlobSidecar{}
+	if err := s.UnmarshalSSZ(encoded); err != nil {
+		return v, err
+	}
+	ro, err := blocks.NewROBlobWithRoot(s, root)
+	if err != nil {
+		return blocks.VerifiedROBlob{}, err
+	}
```

Comment:
Another idea is to just save `blocks.VerifiedROBlob` in DB so we don't need to do this step. We will need to add `VerifiedROBlob` for fastssz. Not sure if it's worth it

Reply:
We do save `VerifiedROBlob` in the db, but remember that from an ssz pov it's just a wrapper around `ethpb.BlobSidecar`. I did consider giving `ROBlob`/`VerifiedROBlob` `UnmarshalSSZ` methods, so they could directly unmarshal ssz values. The only reason they can't do it out of the box is that they need to initialize the internal `*ethpb.BlobSidecar` value to a non-nil instance, so the helper would just initialize the embedded value and then call unmarshal on it.

But I thought it was cleaner to preserve the rule that only the `verification` package gets to create `VerifiedROBlob` (in a non-error path) which we can work on enforcing with a static analyzer. So we will keep something like `verification.BlobSidecarNoop` that *only* the db is allowed to use.